[pytree] update treespec dict keys access

XuehaiPan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #114947
* #116375
* __->__ #116372
* #116374

1. Add a new method `treespec.entries()`
2. Change `treespec.context` access to `treespec.entries()` to hide the implementation details for the `dict` treespec.

cc @zou3519

XuehaiPan committed:
[pytree] update treespec dict keys access

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] update treespec dict keys access"


1. Add a new method `treespec.entries()`
2. Change `treespec.context` access to `treespec.entries()` to hide the implementation details for the `dict` treespec.

cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] update treespec dict keys access"


1. Add a new method `treespec.entries()`
2. Change `treespec.context` access to `treespec.entries()` to hide the implementation details for the `dict` treespec.

cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] update treespec dict keys access"


1. Add a new method `treespec.entries()`
2. Change `treespec.context` access to `treespec.entries()` to hide the implementation details for the `dict` treespec.

cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] update treespec dict keys access"


1. Add a new method `treespec.entries()`
2. Change `treespec.context` access to `treespec.entries()` to hide the implementation details for the `dict` treespec.

cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] update treespec dict keys access"


1. Add a new method `treespec.entries()`
2. Change `treespec.context` access to `treespec.entries()` to hide the implementation details for the `dict` treespec.

cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] update treespec dict keys access"


1. Add a new method `treespec.entries()`
2. Change `treespec.context` access to `treespec.entries()` to hide the implementation details for the `dict` treespec.

cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] update treespec dict keys access"


1. Add a new method `treespec.entries()`
2. Change `treespec.context` access to `treespec.entries()` to hide the implementation details for the `dict` treespec.

cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] update treespec dict keys access"


1. Add a new method `treespec.entries()`
2. Change `treespec.context` access to `treespec.entries()` to hide the implementation details for the `dict` treespec.

cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] update treespec dict keys access"


1. Add a new method `treespec.entries()`
2. Change `treespec.context` access to `treespec.entries()` to hide the implementation details for the `dict` treespec.

cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] update treespec dict keys access"


1. Add a new method `treespec.entries()`
2. Change `treespec.context` access to `treespec.entries()` to hide the implementation details for the `dict` treespec.

cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] update treespec dict keys access"


1. Add a new method `treespec.entries()`
2. Change `treespec.context` access to `treespec.entries()` to hide the implementation details for the `dict` treespec.

cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] update treespec dict keys access"


1. Add a new method `treespec.entries()`
2. Change `treespec.context` access to `treespec.entries()` to hide the implementation details for the `dict` treespec.

cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] update treespec dict keys access"


1. Add a new method `treespec.entries()`
2. Change `treespec.context` access to `treespec.entries()` to hide the implementation details for the `dict` treespec.

cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] update treespec dict keys access"


1. Add a new method `treespec.entries()`
2. Change `treespec.context` access to `treespec.entries()` to hide the implementation details for the `dict` treespec.

cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] update treespec dict keys access"


1. Add a new method `treespec.entries()`
2. Change `treespec.context` access to `treespec.entries()` to hide the implementation details for the `dict` treespec.

cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] update treespec dict keys access"


1. Add a new method `treespec.entries()`
2. Change `treespec.context` access to `treespec.entries()` to hide the implementation details for the `dict` treespec.

cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] update treespec dict keys access"


1. Add a new method `treespec.entries()`
2. Change `treespec.context` access to `treespec.entries()` to hide the implementation details for the `dict` treespec.

cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] update treespec dict keys access"


1. Add a new method `treespec.entries()`
2. Change `treespec.context` access to `treespec.entries()` to hide the implementation details for the `dict` treespec.

cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] update treespec dict keys access"


1. Add a new method `treespec.entries()`
2. Change `treespec.context` access to `treespec.entries()` to hide the implementation details for the `dict` treespec.

cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] update treespec dict keys access"


1. Add a new method `treespec.entries()`
2. Change `treespec.context` access to `treespec.entries()` to hide the implementation details for the `dict` treespec.

cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] update treespec dict keys access"


1. Add a new method `treespec.entries()`
2. Change `treespec.context` access to `treespec.entries()` to hide the implementation details for the `dict` treespec.

cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] update treespec dict keys access"


1. Add a new method `treespec.entries()`
2. Change `treespec.context` access to `treespec.entries()` to hide the implementation details for the `dict` treespec.

cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] update treespec dict keys access"


1. Add a new method `treespec.entries()`
2. Change `treespec.context` access to `treespec.entries()` to hide the implementation details for the `dict` treespec.

cc zou3519

[ghstack-poisoned]

zou3519 commented:
What's the motivation for this?

XuehaiPan commented:
> What's the motivation for this?

@zou3519 This is a prerequisite PR to add this explicit API and hide the implementation details.

- #116375
- #114947

The motivation is to provide an explicit API for accessing the keys of a dict spec. Because the implementation details for `context` may change (i.e. in #114947 the context is changed `original_keys -> (sorted_keys, original_copy)`). This new API allows us to prevent torch and downstream packages from relying on `spec.context` to provide better backward compatibility.

> Also, do you want review on the whole stack?

Each PR in the stack requires the previous PR. Except for the last PR, PRs in the stack are easy to review.

XuehaiPan committed:
Update on "[pytree] update treespec dict keys access"


1. Add a new method `treespec.entries()`
2. Change `treespec.context` access to `treespec.entries()` to hide the implementation details for the `dict` treespec.

cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] update treespec dict keys access"


1. Add a new method `treespec.entries()`
2. Change `treespec.context` access to `treespec.entries()` to hide the implementation details for the `dict` treespec.

cc zou3519

[ghstack-poisoned]

zou3519 commented:
> The motivation is to provide an explicit API for accessing the keys of a dict spec. Because the implementation details for context may change (i.e. in https://github.com/pytorch/pytorch/pull/114947 the context is changed original_keys -> (sorted_keys, original_copy)). This new API allows us to prevent torch and downstream packages from relying on spec.context to provide better backward compatibility.

It might be too late for this, we are able to serialize TreeSpec and that means that the context is a part of the public contract

XuehaiPan committed:
Update on "[pytree] update treespec dict keys access"


1. Add a new method `treespec.entries()`
2. Change `treespec.context` access to `treespec.entries()` to hide the implementation details for the `dict` treespec.

cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] update treespec dict keys access"


1. Add a new method `treespec.entries()`
2. Change `treespec.context` access to `treespec.entries()` to hide the implementation details for the `dict` treespec.

cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] update treespec dict keys access"


1. Add a new method `treespec.entries()`
2. Change `treespec.context` access to `treespec.entries()` to hide the implementation details for the `dict` treespec.

cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] update treespec dict keys access"


1. Add a new method `treespec.entries()`
2. Change `treespec.context` access to `treespec.entries()` to hide the implementation details for the `dict` treespec.

cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]
